[KEP-2400] Mount tmpfs memory-backed volumes with a noswap option if supported

iholder101 opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR addresses https://github.com/kubernetes/kubernetes/issues/105978 which is about preventing memory-backed volumes from being swapped.

This PR brings the following behavior:
* When kubelet starts running with the `fail-swap-on=false` argument, it will check if the tmpfs noswap option is supported. If it's not - a warning log will be raised.
  * To understand if tmpfs noswap is supported or not, kubelet would try to mount a dummy tmpfs with the option and check for success. The directory is obviously cleaned up afterwards.
* If the tmpfs noswap option [1] is supported, use it by default for every tmpfs volume mount. With this options, memory-backed volumes will not be able to swap.

[1] https://www.kernel.org/doc/html/latest/filesystems/tmpfs.html

#### Example
After running kubelet with the `--fail-swap-on=false` argument, I've created the following pod and secret:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mysecret
data:
  username: dXNlcm5hbWU=  # Base64 encoded value of 'username'
  password: cGFzc3dvcmQ=  # Base64 encoded value of 'password'
---
apiVersion: v1
kind: Pod
metadata:
  name: mypod
spec:
  volumes:
  - name: my-volume
    emptyDir:
      medium: "Memory"
  - name: secret-volume
    secret:
      secretName: mysecret
  containers:
  - image: k8s.gcr.io/e2e-test-images/agnhost:2.9
    volumeMounts:
    - name: my-volume
      mountPath: /tmpfsdir
    - name: secret-volume
      mountPath: "/etc/secrets"
    command:
      - sleep
      - "36000"
    name: mycontainer
```

Then on the node we can see tmpfs noswap being used:
```bash
> grep noswap /proc/mounts
tmpfs /var/lib/kubelet/pods/a79290c3-8565-4d05-b065-e6fe3d471beb/volumes/kubernetes.io~projected/kube-api-access-dfw72 tmpfs rw,seclabel,relatime,size=51200k,inode64,noswap 0 0
tmpfs /var/lib/kubelet/pods/1b756da6-531f-40d0-aa44-df79de14ad04/volumes/kubernetes.io~projected/kube-api-access-gh9tk tmpfs rw,seclabel,relatime,size=394492600k,inode64,noswap 0 0
tmpfs /var/lib/kubelet/pods/1fe60428-4258-481a-99b6-bdc4ae671c71/volumes/kubernetes.io~empty-dir/my-volume tmpfs rw,seclabel,relatime,size=524288k,inode64,noswap 0 0
tmpfs /var/lib/kubelet/pods/1fe60428-4258-481a-99b6-bdc4ae671c71/volumes/kubernetes.io~secret/secret-volume tmpfs rw,seclabel,relatime,size=524288k,inode64,noswap 0 0
tmpfs /var/lib/kubelet/pods/1fe60428-4258-481a-99b6-bdc4ae671c71/volumes/kubernetes.io~projected/kube-api-access-sfr4d tmpfs rw,seclabel,relatime,size=524288k,inode64,noswap 0 0
````

#### Which issue(s) this PR fixes:
Fixes #105978

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```

Note: I keep the functest and its revert commit for future reference, as in the future it would be a great candidate for a functional test (when noswap is widely supported enough)

/sig node

iholder101 commented:
/cc @liggitt @kannon92

dims commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -331,10 +331,10 @@ func (ed *emptyDir) setupTmpfs(dir string) error {
 	}
 
-	var options []string
+	options := []string{"noswap"}
```
Looks like 6.4 onwards only!
- https://github.com/torvalds/linux/commit/2c6efe9cf2d7841b75fe38ed1adbd41a90f51ba0
- https://lwn.net/ml/linux-kernel/20230309230545.2930737-7-mcgrof@kernel.org/

liggitt replied:
Questions for node approvers:
1. What other things do we have in the kubelet that assume / require minimum kernel versions?
2. How do we detect and document those minimum requirements?
3. What do we do when the kernel doesn't meet those requirements? (error early on startup, error late at runtime, non-error fail open at runtime, something else?)

iholder101 replied:
Thanks a lot @liggitt and @dims!

That's very helpful. I'll definitely look into it.

harche replied:
>What other things do we have in the kubelet that assume / require minimum kernel versions?

When using a metric `container_memory_max_usage_bytes` in cgroup v2, the minimum kernel required kernel is 5.19, otherwise that metric will return 0.

xref: https://github.com/kubernetes/kubernetes/pull/122664#issuecomment-1899581854

liggitt replied:
A minimum kernel version to have specific metrics available is a good data point. Do we have examples where a minimum kernel version is required for a container or volume to operate as expected?

haircommander replied:
minimum kernel version is tricky though because downstreams backport patches to older kernel versions. having a way to probe the kernel for a specific feature is more reliable (like doing a dummy mount at kubelet startup to see if tmpfs supports noswap)

liggitt replied:
Probing at start sounds great if we can make it reliable. I guess we would only need to do this if swap was enabled.

The next question is what the kubelet should do if swap is enabled, and the kubelet is configured to run with swap on, and the kernel version doesn't allow `noswap` for tmpfs. I *think* the kubelet should fail safe and refuse to run.

mrunalp replied:
> The next question is what the kubelet should do if swap is enabled, and the kubelet is configured to run with swap on, and the kernel version doesn't allow noswap for tmpfs. I think the kubelet should fail safe and refuse to run.

Yes, that sounds good.

kannon92 replied:
It may be difficult to enforce that last point as it seems that developer environments (hack/local-up-cluster enables swap by default) may use swap. Since tmpfs swap is the default option, I think that limits our options.

I know that when I tried to fail swap on cgroup v1 I ran into quite a few tests/ci environments that failed on that which was surprising to me.

I think "swap is enabled" will have to mean `--fail-swap-on=false` in this case as this is not really limited to pods swapping but whether or not kubelet won't fail on swap.

At a quick search, I did see some k8s components that set to this to false. https://cs.k8s.io/?q=failSwapOn%3A%20false&i=nope&files=&excludeFiles=&repos=

kannon92 replied:
Looking at these test logs..

```
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074116     745 mount_linux.go:224] Mounting cmd (mount) with arguments (-t tmpfs -o noswap,size=54736998400 tmpfs /var/lib/kubelet/pods/75271a50-bac8-4506-8961-cf8331019a18/volumes/kubernetes.io~projected/kube-api-access-tvkpq)
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074051     745 empty_dir_linux.go:99] Statfs_t of /var/lib/kubelet/pods/45452852-ccfd-4739-8693-6e216c5f31cb/volumes/kubernetes.io~projected/kube-api-access-pzjxq: {Type:61267 Bsize:4096 Blocks:193176694 Bfree:185170145 Bavail:175338952 Files:49135616 Ffree:48642004 Fsid:{Val:[1454556786 880505981]} Namelen:255 Frsize:4096 Flags:4128 Spare:[0 0 0 0]}
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074213     745 empty_dir.go:340] pod 45452852-ccfd-4739-8693-6e216c5f31cb: mounting tmpfs for volume wrapped_kube-api-access-pzjxq
Mar 26 15:10:10 kind-control-plane kubelet[745]: I0326 15:10:10.074234     745 mount_linux.go:224] Mounting cmd (mount) with arguments (-t tmpfs -o noswap,size=52428800 tmpfs /var/lib/kubelet/pods/45452852-ccfd-4739-8693-6e216c5f31cb/volumes/kubernetes.io~projected/kube-api-access-pzjxq)
Mar 26 15:10:10 kind-control-plane kubelet[745]: E0326 15:10:10.081410     745 mount_linux.go:236] Mount failed: exit status 32
```

Looks like we got the behavior we want if the feature is not set. We get an exit code of 32 when mounting on startup. Kind uses fail-swap-on=false by default I guess.

iholder101 replied:
@kannon92 

> Looks like we got the behavior we want if the feature is not set.

Yes, that's true, in the current implementation the `noswap` option is always being added, not only when using `--fail-swap-on=false`. However I agree with @liggitt that it sounds like the right path forward. I'll work on that.

> We get an exit code of 32 when mounting on startup. 

Do we know which kernel version is used in k8s CI? Is it >=6.4?

> Probing at start sounds great if we can make it reliable.

I fully agree. I'll try to understand how / if it's impossible to do so.

kannon92 replied:
> Do we know which kernel version is used in k8s CI? Is it >=6.4?

Its difficult to answer that question. 

On the pull-e2e-kind tests, I see `Mar 26 15:09:49 kind-control-plane kubelet[250]: I0326 15:09:49.330463     250 manager.go:233] Version: {KernelVersion:5.15.0-1048-gke ContainerOsVersion:Debian GNU/Linux 12 (bookworm) DockerVersion: DockerAPIVersion: CadvisorVersion: CadvisorRevision:}
`

So I can see that we using debian 12 and supposely the kernel is 5.15? But I think there are backports. @haircommander mentioned to me that using kernel version is pretty much impossible with all these back ports.

kannon92 replied:
@pacoxu has this issue https://github.com/kubernetes/kubernetes/issues/116799 where they mention features and kernel support.

Irvingwangjr replied:
Is there any chance that we make this as a mount opts? I know this is KEP focusing on swap thing, but we might hope to custumize more opts like huge=advise for THP, or readOnly kind of opts

liggitt replied:
I'm not sure if you're proposing exposing mount options via the API, but those are not generally safe to expose to end users. The intent of this is not to expose more levers and knobs, but to make the existing API surface continue to operate safely even when swap is enabled.

Irvingwangjr replied:
yes, I'm proposing exposing mount Options. But also agree with your concern here, like noexec might bring unexpected security issue

Irvingwangjr replied:
but how about making a subset as mount Options?

kannon92 commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -331,10 +331,10 @@ func (ed *emptyDir) setupTmpfs(dir string) error {
 	// Linux system default is 50% of capacity.
 	if ed.sizeLimit != nil && ed.sizeLimit.Value() > 0 {
-		options = []string{fmt.Sprintf("size=%d", ed.sizeLimit.Value())}
+		options = append(options, fmt.Sprintf("size=%d", ed.sizeLimit.Value()))
```
Please add some unit tests verifying the noswap option.

I have a feeling this code will change but let's make sure to have some unit test coverage also.

iholder101 replied:
Done!

bart0sh commented:
/triage accepted
/priority important-longterm

iholder101 committed:
unit test: Use tmpfs noswap if supported

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
Thanks a lot everyone for the review and discussion.

Following the discussion we had on the corresponding [issue](https://github.com/kubernetes/kubernetes/issues/105978) I'm now taking a slightly different approach. @liggitt [summarized](https://github.com/kubernetes/kubernetes/issues/105978#issuecomment-2078246497) our discussion elegantly with the following table:

<table>
<tr>
 <td>
 <td>swap off
 <td>swap on, supports noswap
 <td>swap on, doesn't support noswap
<tr>
 <td>--fail-swap-on unset
 <td>-
 <td>fail on kubelet start
 <td>fail on kubelet start
<tr>
 <td>--fail-swap-on=true
 <td>-
 <td>fail on kubelet start
 <td>fail on kubelet start
<tr>
 <td>--fail-swap-on=false
 <td>-
 <td>detect on start, use noswap
 <td>detect on start, warn, don't use noswap
</table>

For more info please refer to the discussion on the corresponding issue.

TLDR:
* The `--fail-swap-on` argument was not presented by the NodeSwap feature and predated it.
* This approach keeps backward compatibility while improving the situation.
* From the NodeSwap feature perspective, this should be enough to fulfil the KEP's requirements.

haircommander commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -559,3 +556,16 @@ func getVolumeSource(spec *volume.Spec) (*v1.EmptyDirVolumeSource, bool) {
+		options = append(options, fmt.Sprintf("size=%d", ed.sizeLimit.Value()))
+	}
+
+	if noswapSupportd {
```
nit: noswapSupported*?

iholder101 replied:
Thank you! fixed

haircommander commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +66,47 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+)
+
+func IsTmpfsNoswapOptionSupported(mounter mount.Interface) bool {
+	isTmpfsNoswapOptionSupportedHelper := func() bool {
```
why define a functor? you can just put the contents of it into the Do func

iholder101 replied:
I think it's more elegant this way, as the function is very cohesive and focuses on answering the actual question of whether tmpfs noswap is supported or not. Otherwise I'd have to both identify if it's supported and assign the variable.

So it would look something like:
```go
if err != nil {
	klog.InfoS("error creating dir ...")
	tmpfsNoswapOptionSupported = false
	return
}
```

It's both unelegant and error prone IMO, for example it's easy to forget to return after the assignment is made.

But that's just one man's opinion :) If people think it's more elegant the other way I'm willing to make the change :+1:

haircommander replied:
fair enough, this works for me :)

rphillips commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +66,47 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+			return false
+		}
+
+		mountDir := filepath.Join("/", "tmp", "tmpfs-noswap-test")
```
os.MkdirTemp would probably be a better fit here.

iholder101 replied:
Done, thanks!

iholder101 commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +66,47 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+			klog.ErrorS(err, "error removing test tmpfs dir", "mount path", mountDir)
+		}
+
+		err = mounter.MountSensitiveWithoutSystemd("tmpfs", mountDir, "tmpfs", []string{"noswap"}, nil)
```
Thank you!
I've made the `"noswap"` string a const, since it's being used more than the slice.

kwilczynski commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +65,50 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+			return false
+		}
+
+		mountDir, err := os.MkdirTemp("", "tmpfs-noswap-test")
```
If you add the dash at the end, then a random suffix will be automatically appended. Might help in some cases where things have failed and you have an orphaned directory left behind (which is not a problem since `/tmp` tends to be cleaned up on startup or in some cases even periodically).

```suggestion
		mountDir, err := os.MkdirTemp("", "tmpfs-noswap-test-")
```

kwilczynski commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +65,50 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+
+		mountDir, err := os.MkdirTemp("", "tmpfs-noswap-test")
+		if err != nil {
+			klog.InfoS("error creating dir to test if tmpfs noswap is enabled. Assuming not supported", "mount path", mountDir, "error", err)
```
Do we want "info" level here for this error? Perhaps a warning or a real error log level?

iholder101 replied:
I was initially using `klog.Warning()`, but the linter complained that I must use structured logging here (i.e. something that ends with an S, `InfoS`, `ErrorS`, etc).

Unfortunately (not sure why) there is no `WarningS`.

But using an `ErrorS` is definitely an option. I chose not to use it since it's not an error in the sense that kubelet cannot start/run properly, but rather a best-effort attempt to use noswap.

kwilczynski replied:
I see. This is a bit unfortunate and an odd choice not to include this warning print variant.

kwilczynski commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +65,50 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+
+		err = mounter.MountSensitiveWithoutSystemd("tmpfs", mountDir, "tmpfs", []string{TmpfsNoswapOption}, nil)
+		if err != nil {
+			klog.InfoS("error mounting tmpfs with the noswap option. Assuming not supported", "error", err)
```
Ditto. As the above comment.

iholder101 commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -224,6 +225,10 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 		}
+
+		if !util.IsTmpfsNoswapOptionSupported(mountUtil) {
+			klog.InfoS("tmpfs noswap option is not supported, hence memory-backed volumes (e.g. secrets, emptyDirs) might be swapped to disk")
```
> I think this will warn on machines where swap is not even enabled

I think this behavior is desired.

If somebody uses failSwapOn + disabled NodeSwap then secrets can still be swapped to disk if a swap file/partition exists on the node.

This bug/security risk existed from whenever the failSwapOn was introduced, before NodeSwap even existed. IOW, this problem is out-of-scope for the NodeSwap feature, but it's rather a `--fail-swap-on` bug.

liggitt replied:
I'm not talking about the NodeSwap feature gate, I'm talking about swap in the OS… if that's not enabled, tmpfs volumes can't swap out, right?

iholder101 replied:
Aha, got you.
By "swap enabled in the OS" do you mean that there's an existing swap file/partition? This makes total sense to me :+1:

liggitt replied:
yeah, whatever we're using above to detect swap is enabled so the kubelet can fail if `failSwapOn` is true

iholder101 replied:
Thanks for your reviews and clarifications! sounds great to me.
Let me rework the code and let you know when it's ready to review again :+1:

iholder101 replied:
@liggitt the code is refactored now and is much clearer IMO.
I believe I've addressed your feedback here.

PTAL

iholder101 commented on pkg/kubelet/util/util.go:
```diff
@@ -60,3 +65,50 @@ func GetContainerByIndex(containers []v1.Container, statuses []v1.ContainerStatu
+			return false
+		}
+
+		mountDir, err := os.MkdirTemp("", "tmpfs-noswap-test")
```
Good point.
@kwilczynski maybe you can help answering that?

kwilczynski replied:
There are two things to address here:

- Does Kubernetes have a known, perhaps even dedicated, location somewhere on the file system it can use?
- Would there always be access to the common temporary directory for Kubernetes?

The answer to the first question would be: I am unaware of such a directory or a mount point that Kubernetes creates and uses exclusively for its own needs. Usually, the role of such location somewhere on the file system would have been filed by the common temporary directory. However, perhaps Kubernetes should provide this under, let's say, `/run/kubelet`? or `/run/kubernetes`? or perhaps `/run/k8s`? This is where temporary files, directories, tests of sorts, tag files, etc., could be stored. This would also make it easier to make SELinux or AppArmor policies, and such, should there be a need for such a Kubernetes-curated location.

Second question: Using a common system-wide shared temporary directory is something that, on most Linux systems, would have been taken for granted and provided. Any process, whether running as a privileged user or not, should be able to save some temporary data it needs on the side there. We have /tmp and /var/tmp as being the two most commonly used locations, and a lot of libraries tend to hard code at least one of these locations (/tmp most commonly since it's ephemeral in terms of how its content lifespan looks like). Even with systemd and private temporary directory support enabled, a process should have no issues writing to its assumed default location.

Now, whether Kubernetes could be restricted from using /tmp or /var/tmp, of course it could, as it would be something that the user has done on purpose, most of the time, since by default, the expectation is that writes to files or directories under /tmp will succeed.

Having said all that, the default temporary directory location would be a perfectly reasonable choice for Kubernetes. Should it fail, then we are dealing with something non-standard on the underlying node, or within some runtime environment, which would be something different to look into and the user will be most likely aware of that, hopefully.

See the following for more details about temporary directories:

- [Using /tmp/ and /var/tmp/ Safely](https://systemd.io/TEMPORARY_DIRECTORIES)
- [file-hierarchy — File system hierarchy overview](https://www.freedesktop.org/software/systemd/man/latest/file-hierarchy.html)

liggitt replied:
> * Does Kubernetes have a known, perhaps even dedicated, location somewhere on the file system it can use?
> * Would there always be access to the common temporary directory for Kubernetes?
> 
> The answer to the first question would be: I am unaware of such a directory or a mount point that Kubernetes creates and uses exclusively for its own needs.

isn't that the kubelet RootDirectory?

https://github.com/kubernetes/kubernetes/blob/7a6931b31cf04d2a91ce8c0af09ad0f474927436/pkg/kubelet/kubelet.go#L1380-L1430

kwilczynski replied:
@liggitt, that would work. Given the following code comment:

```go
// getRootDir returns the full path to the directory under which kubelet can
// store data.  These functions are useful to pass interfaces to other modules
// that may need to know where to write data without getting a whole kubelet
// instance.
func (kl *Kubelet) getRootDir() string {
	return kl.rootDirectory
}
```

Can we use it to write anything? Or are there some expectations of the type of data being written there? If not, then it would work fine indeed.

@iholder101, thoughts?

iholder101 replied:
Thanks a lot for this discussion!
This is implemented in [`e8b303d` (#124060)](https://github.com/kubernetes/kubernetes/pull/124060/commits/e8b303d5da6d20a6628bbd7e35eebfe3fe2cd658).

liggitt commented on pkg/volume/emptydir/empty_dir.go:
```diff
@@ -331,11 +332,7 @@ func (ed *emptyDir) setupTmpfs(dir string) error {
-	if ed.sizeLimit != nil && ed.sizeLimit.Value() > 0 {
-		options = []string{fmt.Sprintf("size=%d", ed.sizeLimit.Value())}
-	}
+	options := ed.generateTmpfsMountOptions(util.IsTmpfsNoswapOptionSupported(ed.mounter))
```
should we only use `noswap` if swap is enabled in the OS?

iholder101 replied:
Definitely. Already working on that.

iholder101 replied:
@liggitt After finishing the implementation - I'm second guessing this statement :)

One corner case is if a user would start kubelet and only then create a swap file. kubelet currently does not detect if that's happening. The admin can restart kubelet, but then I guess the already mounted volumes would lack the `noswap` option.

I'm not sure if there's any reason not to use the `noswap` option if it's supported. I even think that in the future, when it is widely supported enough, we can drop this logic and use the `tmpfs` option indefinitely.

WDYT?

antoinedeschenes replied:
> @liggitt After finishing the implementation - I'm second guessing this statement :)
> 
> One corner case is if a user would start kubelet and only then create a swap file. kubelet currently does not detect if that's happening. The admin can restart kubelet, but then I guess the already mounted volumes would lack the `noswap` option.
> 
> I'm not sure if there's any reason not to use the `noswap` option if it's supported. I even think that in the future, when it is widely supported enough, we can drop this logic and use the `tmpfs` option indefinitely.
> 
> WDYT?

Indeed, what if you attach a volume, create a swap partition and mount it through a DaemonSet?

liggitt replied:
I don't know enough about the mount option to know if there's any downside / risk / side-effect to using it.

If we're 100% certain it's a no-op when swap is not enabled, that seems ok.

If not, I'd lean towards not changing behavior of volumes on hosts without swap, at least not in the first release.

I'll defer to node / storage folks who know the implications of the option better.

iholder101 replied:
FYI @mrunalp @SergeyKanzhelev

kwilczynski replied:
Of course, I forgot about old kernels... so this _would_ fail for a kernel that is too old to support "noswap" as a mount option, per:

```console
root@vagrant:~# strace -s 65535 -e mount -- mount -t tmpfs -o noswap tmpfs /mnt
mount("tmpfs", "/mnt", "tmpfs", 0, "noswap") = -1 EINVAL (Invalid argument)
mount: /var/mnt: wrong fs type, bad option, bad superblock on tmpfs, missing codepage or helper program, or other error.
+++ exited with 32 +++
```

So we still would need to have the active tmpfs mount check, just in case...

iholder101 replied:
Until there are objections from anyone, the implementation remains to try to always use tmpfs if supported.

> So we still would need to have the active tmpfs mount check, just in case...

Indeed. The check is already in place, and we should drop it in the future IMO

iholder101 committed:
Use tmpfs noswap if supported

use the tmpfs noswap option in order
to mount memory-backed volumes if it's supported.

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Warn of swap is enabled on the OS and tmpfs noswap is not supported

When --fail-swap-on=false kubelet CLI argument
is provided, but tmpfs noswap is not supported
by the kernel, warn about the risks of memory-backed
volumes being swapped into disk

Signed-off-by: Itamar Holder <iholder@redhat.com>

kwilczynski commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,109 @@
+		swapData, err := os.ReadFile(swapFilePath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				klog.InfoS("File does not exist, assuming that swap is disabled", "path", swapFilePath)
```
The procfs'  `/proc/swaps` file will always be there as long as `CONFIG_PROC_FS` is enabled, and this is usually true. There might be an issue reading the file, but that does not mean swap is disabled.

iholder101 replied:
> There might be an issue reading the file, but that does not mean swap is disabled.

I agree. The question is what we should do in such a scenario.
I've followed the previous implementation (before this PR), which assumed swap is not enabled if the file is not there.

Would you suggest to change this behavior? What do you have in mind?

kwilczynski replied:
> > There might be an issue reading the file, but that does not mean swap is disabled.
> 
> I agree. The question is what we should do in such a scenario. I've followed the previous implementation (before this PR), which assumed swap is not enabled if the file is not there.

No worries. Makes sense to keep any prior art.
 
> Would you suggest to change this behavior? What do you have in mind?

This code in the kernel is almost 20 years old, and I have never seen this profs object not being there while procfs itself is enabled. Thus, I am not sure if checking for this file presence explicitly makes a lot of sense?

![Screenshot from 2024-05-09 17-47-01](https://github.com/kubernetes/kubernetes/assets/250723/142eb073-158b-4773-af79-8b55d47c9638)

However, it is probably best to leave this code as-is, if this was a lift from an existing implementation. :smile:

iholder101 replied:
:)
Let's defer this as well. I'm open for discussing removing it in the future :+1:

kwilczynski commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,109 @@
+		const swapFilePath = "/proc/swaps"
+		swapData, err := os.ReadFile(swapFilePath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				klog.InfoS("File does not exist, assuming that swap is disabled", "path", swapFilePath)
+				return false, nil
+			} else {
+				return false, err
+			}
```
There is no need for an if-else statement, since there is only a `return` in the code block above.

```suggestion
			if os.IsNotExist(err) {
				klog.InfoS("File does not exist, assuming that swap is disabled", "path", swapFilePath)
				return false, nil
			}
			return false, err
```

iholder101 replied:
Thank you! Done!

kwilczynski commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		const swapFilePath = "/proc/swaps"
+		swapData, err := os.ReadFile(swapFilePath)
```
So, `os.ReadFile()` internally calls `fstat(2)` in order to get the size of the file it's reading. For a lot of procfs and sysfs (kernfs) objects, the size will be `0` anyway. As such, you can avoid this extra call with a combination of `os.Open()` and then reading a page (4 KiB) worth of data in a single go, for example:

```go
f, err := os.Open(file)
if err != nil {
	return err
}
defer f.Close()

data, err := io.ReadAll(io.LimitReader(f, 4096))
if err != nil {
	return err
}
```

Why page worth of data? The amount of data that procfs or sysfs can store and that can be read in one go is 4 KiB, a default page size (for the sake of simplicity, we assume that there isn't any difference between architectures) on most platforms.

This is not something that must need to be changed since this function is run once and during the kubelet start-up. More of a recommendation or advice.

iholder101 replied:
Thanks a lot @kwilczynski, makes total sense.
However TBH I prefer to keep this PR focused as much as possible, and I feel changing this behavior is a bit out-of-scope.

I can definitely submit a PR for improving that as a follow-up. Does that sound reasonable?

kwilczynski replied:
@iholder101, oh no worries. With you there.

Like I said, this is just a "by the way..." kind of comment, for dealing with that sort of pseudo file systems, etc.

iholder101 replied:
Cool :)
I'll cc you at the follow-up PR :+1:

yujuhong replied:
That looks like a optimization, and is probably better suited to be factored out as a separate function for reusability. :)

kwilczynski commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,109 @@
+
+		baseMountDir := rootDir
+		if baseMountDir == "" {
+			klog.ErrorS(fmt.Errorf("swap utils root dir is not initialized, falling back to /tmp"), "")
```
```suggestion
			klog.ErrorS(errors.New("swap utils root dir is not initialized, falling back to /tmp"), "")
```

iholder101 replied:
Out of curiosity, what's the difference between the two? Why is your suggestion better?

kwilczynski replied:
A rule of thumb: don't have any format strings to process, then don't use a function that would look for them. Some linters also pick on this. Also, `errors.New()` will be somewhat slightly faster, since no need to check if there is anything to parse, etc.

kwilczynski commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,25 +203,19 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-		} else {
-			swapData = bytes.TrimSpace(swapData) // extra trailing \n
-			swapLines := strings.Split(string(swapData), "\n")
+	swap.InitializeRootDir(nodeConfig.KubeletRootDir)
```
Nice! :+1:

kwilczynski commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,125 @@
+	rootDir = path
+}
+
+func IsTmpfsNoswapOptionSupported(mounter mount.Interface) bool {
```
There is a potential optimisation you can do here to avoid having to do the mount test.

You can check if the kernel is > 6.4, and if so, short-circuit and return `true` since the feature will most definitely be there. But then, for anything else, you can do the mount attempt to verify things 100% (and support older kernels with backported changes).

We already do this to discover whether some kernel features are supported, per:

- [pkg/util/kernel/version.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/kernel/version.go)
- [pkg/util/kernel/constants.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/util/kernel/constants.go)
- [pkg/proxy/ipvs/proxier.go#L299-L314](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L299-L314)

Thoughts?

iholder101 replied:
Awesome! Thanks a lot, I've added that

haircommander commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,25 +203,19 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
+		}
+
+		if !swap.IsTmpfsNoswapOptionSupported(mountUtil) {
+			klog.InfoS("tmpfs noswap option is not supported, hence memory-backed volumes (e.g. secrets, emptyDirs) might be swapped to disk")
```
In my ideal kubelet would post an event or metric so this state is more visable to an admin than just a log line. WDYT? It doesn't necessarily have to block this and could be a follow-up I think

yujuhong commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,125 @@
+
+// IsSwapOn detects whether swap in enabled on the system by inspecting
+// /proc/swaps. If the file does not exist, an os.NotFound error will be returned.
+// If running on windows, swap is assumed to always be false.
```
Could we add the expected content of /proc/swaps here for readability, or better yet, with a unit test?

iholder101 committed:
Replace log entry by a warning event

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
Add a isSwapOnAccordingToProcSwaps() function and swap utils unit tests

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 committed:
If the kernel version is at least 6.4, assume tmpfs noswap is supported

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
@kwilczynski @haircommander @yujuhong 

I believe I've addressed all of your comments above. Thanks a lot for your reviews!
Can you PTAL?

iholder101 commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

kwilczynski commented:
> I believe I've addressed all of your comments above. Thanks a lot for your reviews! Can you PTAL?

@iholder101, thank you! Looks very good! :tada: 

/approve
/lgtm

haircommander commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,25 +204,21 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-			}
+	if isSwapOn {
+		if failSwapOn {
+			return nil, fmt.Errorf("running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false")
```
nit: is the `!` intentional after `swap`?

kwilczynski replied:
I took it as two sentences being there, with the first one being very assertive :smile:

haircommander replied:
if that's intentional then the nit would be capitalize Or (though I think I prefer them as the same sentence)

kwilczynski replied:
Yeah. It might have been kept all lower-case since this is an error message.

The entire message could, perhaps, be improved too... Hmm.

iholder101 replied:
This message preexisted this PR actually, I've just moved it:
https://github.com/kubernetes/kubernetes/blob/8362f51c6620a55953ac564d24fa82d2f2ea2403/pkg/kubelet/cm/container_manager_linux.go#L224

But yeah, this `!` seems out of place, I'll remove it

kwilczynski commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,25 +204,21 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
+		}
+
+		if !swap.IsTmpfsNoswapOptionSupported(mountUtil) {
+			const msg = "tmpfs noswap option is not supported, hence memory-backed volumes (e.g. secrets, emptyDirs) might be swapped to disk"
+			nodeRef := nodeRefFromNode(string(nodeConfig.NodeName))
+			recorder.Event(nodeRef, v1.EventTypeWarning, events.PossibleMemoryBackedVolumesOnDisk, msg)
```
A small nit. Some improvement to the event message, and to potentially drop the constant.

```suggestion
			nodeRef := nodeRefFromNode(string(nodeConfig.NodeName))
			recorder.Event(nodeRef, v1.EventTypeWarning, events.PossibleMemoryBackedVolumesOnDisk,
				"The tmpfs noswap option is not supported. Memory-backed volumes (e.g. secrets, emptyDirs, etc.) "+
					"might be swapped to disk and should no longer be considered secure.",
	)
)
```

haircommander commented:
/lgtm

thanks for your work here!

yujuhong commented:
> @kwilczynski @haircommander @yujuhong
> 
> I believe I've addressed all of your comments above. Thanks a lot for your reviews! Can you PTAL?

Thanks! I think other reviewers here are more familiar with the code touched in the PR, but looks good to me on the high-level :) 

/lgtm

dims commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -20,12 +20,12 @@ limitations under the License.
-	"bytes"
 	"context"
 	"fmt"
+	"k8s.io/kubernetes/pkg/kubelet/events"
```
these 2 lines should be in the correct section (see the list of imports from line 39 through 68). these 2 lines belong there.

dims commented on pkg/kubelet/util/swap/swap_util.go:
```diff
@@ -0,0 +1,143 @@
+import (
+	"bytes"
+	"errors"
+	"k8s.io/apimachinery/pkg/util/version"
```
see the pattern of how the imports are organized in pkg/kubelet/cm/container_manager_linux.go, please follow the same here (avoid mixing all the imports into a single section)

dims commented on pkg/volume/emptydir/empty_dir_test.go:
```diff
@@ -21,8 +21,10 @@ package emptydir
 
 import (
 	"fmt"
+	"k8s.io/kubernetes/pkg/kubelet/util/swap"
```
same feedback about putting imports in the sections they belong

iholder101 commented:
Thanks @dims, done.
PTAL.

dims commented:
/approve
/hold

( for explicit `/lgtm` from either @liggitt or @mrunalp )

kannon92 commented:
Nice work @iholder101!

can you update the swap KEP with some of our discoveries around this?

liggitt commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,25 +204,23 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-		} else {
-			swapData = bytes.TrimSpace(swapData) // extra trailing \n
-			swapLines := strings.Split(string(swapData), "\n")
+	swap.InitializeRootDir(nodeConfig.KubeletRootDir)
```
I'm not a fan of having to set a global for IsTmpfsNoswapOptionSupported to work properly

can `IsTmpfsNoswapOptionSupported` take a writeable tmp dir to use? we could feed it here from nodeConfig.KubeletRootDir and from setupTmpfs with `ed.plugin.host.GetPluginDir(emptyDirPluginName)` or something similar

iholder101 replied:
Done!

BTW, the reasoning for what I did is that the new `mountPath` parameter added to `IsTmpfsNoswapOptionSupported` will practically be used only at the first use, which will always be called from `container_manager_linux.go` with `KubeletRootDir`. In other words, `ed.plugin.host.GetPluginDir(emptyDirPluginName)` will never be practically used for mounting.

That being said, I think your concern is valid. Code can change and move around and we better ensure it's always safe to use.

Thanks @liggitt! PTAL

liggitt commented:
> ( for explicit `/lgtm` from either @liggitt or @mrunalp )

one comment about avoiding the global rootDir in the swap utility package, I'll defer to @mrunalp for lgtm after that's addressed

mrunalp commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -212,7 +212,7 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 	if isSwapOn {
 		if failSwapOn {
-			return nil, fmt.Errorf("running with swap on is not supported, please disable swap! or set --fail-swap-on flag to false")
+			return nil, fmt.Errorf("running with swap on is not supported, please disable swap or set --fail-swap-on flag to false")
```
Please fold this into the commit where this is first introduced.

iholder101 replied:
The reasoning for having this in a separate commit is that as [said above](https://github.com/kubernetes/kubernetes/pull/124060#discussion_r1601753544) this error message predated this PR, so in one commit I was just moving it around, and in another I was improving it. I thought it would be easier to understand this way and make every commit more concise.

In any case, this commit is now squashed. :+1:

mrunalp commented:
@iholder101 Can we make the commits purely additive?

iholder101 committed:
Use kubelet owned directories for mounting rather than /tmp

Signed-off-by: Itamar Holder <iholder@redhat.com>

iholder101 commented:
> @iholder101 Can we make the commits purely additive?

@mrunalp Definitely.
I've now folded two commits (regarding error message improvement + address dim's request to organize imports) into previous ones.

PTAL

liggitt commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,25 +204,22 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-			swapLines := strings.Split(string(swapData), "\n")
+	isSwapOn, err := swap.IsSwapOn()
+	if err != nil {
+		return nil, fmt.Errorf("failed to determine if swap is on: %w", err)
```
is this a new error condition? there was no swap detection error that hard-failed before

haircommander commented:
/lgtm

great work here!

AkihiroSuda commented:
Regression:
- https://github.com/kubernetes/kubernetes/issues/125137
